src/command/list/json.rs

```diff
@@ -0,0 +1,420 @@
+use serde_json::to_string_pretty;
+
+use super::{Node, Package, PackageManager, Toolchain};
+
+pub(super) fn format(toolchain: &Toolchain) -> Option<String> {
+    let (runtimes, package_managers, packages) = match toolchain {
+        Toolchain::Node(runtimes) => (describe_runtimes(&runtimes), None, None),
+        Toolchain::PackageManagers(package_managers) => {
+            (None, describe_package_managers(&package_managers), None)
+        }
+        Toolchain::Packages(packages) => (None, None, describe_packages(&packages)),
+        Toolchain::Tool {
+            name,
+            host_packages,
+        } => (None, None, Some(describe_tool_set(name, host_packages))),
+        Toolchain::Active {
+            runtime,
+            package_manager,
+            packages,
+        } => (
+            runtime
+                .as_ref()
+                .and_then(|r| describe_runtimes(&[(**r).clone()])),
+            package_manager
+                .as_ref()
+                .and_then(|p| describe_package_managers(&[(**p).clone()])),
+            describe_packages(&packages),
+        ),
+        Toolchain::All {
+            runtimes,
+            package_managers,
+            packages,
+        } => (
+            describe_runtimes(&runtimes),
+            describe_package_managers(&package_managers),
+            describe_packages(&packages),
+        ),
+    };
+
+    match (runtimes, package_managers, packages) {
+        (Some(runtimes), Some(package_managers), Some(packages)) => {
+            Some(format!("{},{},{}", runtimes, package_managers, packages))
+        }
+        (Some(runtimes), Some(package_managers), None) => {
+            Some(format!("{},{}", runtimes, package_managers))
+        }
+        (Some(runtimes), None, Some(packages)) => Some(format!("{},{}", runtimes, packages)),
+        (Some(runtimes), None, None) => Some(format!("{}", runtimes)),
+        (None, Some(package_managers), Some(packages)) => {
+            Some(format!("{},{}", package_managers, packages))
+        }
+        (None, Some(package_managers), None) => Some(format!("{}", package_managers)),
+        (None, None, Some(packages)) => Some(format!("{}", packages)),
+        (None, None, None) => None,
+    }
+}
+
+fn describe_runtimes(runtimes: &[Node]) -> Option<String> {
+    #[derive(serde::Serialize)]
+    struct Runtimes<'a> {
+        runtimes: &'a [Node],
+    };
+    if runtimes.is_empty() {
+        None
+    } else {
+        Some(to_string_pretty(&Runtimes { runtimes }).unwrap())
```

Comment:
One suggestion as I'm looking at this: I think the *default* value for JSON printing should be a compact representation, and we should add a `--pretty` flag when printing JSON that uses `to_string_pretty` instead of `to_string`. What do you think @MusiKid and @charlespierce?

Reply:
I think that it's indeed a better idea to provide an option for the pretty output. How we should implement it however? We can create two different `format` function, one for the pretty output and another other for the compact one, or provide a new argument for the function like this:
```rust
pub(super) fn format(toolchain: &Toolchain, pretty: bool) -> Option<String>
```
What do you think?

Reply:
Sorry for the long delay in response! I'll try to get back to this in the next week or so!

Reply:
OK no problem!